File: template/Template.py
from typing import Protocol, Optional

class WorkerTemplate(Protocol):
    def start_work(cls) -> None: 
        print("Starting work")
    
    def stop_work(cls) -> None:
        print("Stopping work")

    def do_work(cls) -> None:
        pass

    def execute(self) -> None:
        self.start_work()
        self.do_work()
        self.stop_work()
        
class HRWorker(WorkerTemplate):
    def do_work(self) -> None:
        print("HR worker is working")

class ITWorker(WorkerTemplate):
    def do_work(self) -> None:
        print("IT worker is working")

class Manager(WorkerTemplate):
    def do_work(self) -> None:
        print("Manager is working")

if __name__ == '__main__':
    work: WorkerTemplate = HRWorker()
    work.execute()

    work = ITWorker()
    work.execute()

    work = Manager()
    work.execute()

File: template/Template.cpp
#include <iostream>
#include <string>

using namespace std;

class WorkerTemplate
{
public:
    void execute()
    {
        startWork();
        doWork();
        endWork();
    }
    void startWork()
    {
        cout << "Start work" << endl;
    }
    void endWork()
    {
        cout << "End work" << endl;
    }

    virtual void doWork() = 0;
    virtual ~WorkerTemplate() {}
};

class HRWorker : public WorkerTemplate
{
public:
    void doWork()
    {
        cout << "HR worker do work" << endl;
    }
};

class ManagerWorker : public WorkerTemplate
{
public:
    void doWork()
    {
        cout << "Manager worker is managing task" << endl;
    }
};

class DeveloperWorker : public WorkerTemplate
{
public:
    void doWork()
    {
        cout << "Developer worker is writing code" << endl;
    }
};

// as a client
int main(int argc, char *argv[])
{
    WorkerTemplate *worker = new HRWorker();
    worker->execute();
    delete worker;

    worker = new ManagerWorker();
    worker->execute();
    delete worker;

    worker = new DeveloperWorker();
    worker->execute();
    delete worker;
    return 0;
}